src/collation.rs

```diff
@@ -48,3 +52,235 @@ pub struct Collation {
+    /// "at").
+    ///
+    /// This is also called the level-2 strength.
+    Secondary,
+
+    /// Upper and lower case differences in characters are distinguished at the tertiary level (for
+    /// example, "ao" < "Ao" < "aò").
+    ///
+    /// This is also called the level-3 strength.
+    Tertiary,
+
+    /// When punctuation is ignored at level 1-3, an additional level can be used to distinguish
+    /// words with and without punctuation (for example, "ab" < "a-b" < "aB").
+    ///
+    /// This is also called the level-4 strength.
+    Quaternary,
+
+    /// When all other levels are equal, the identical level is used as a tiebreaker. The Unicode
+    /// code point values of the NFD form of each string are compared at this level, just in
+    /// case there is no difference at levels 1-4.
+    ///
+    /// This is also called the level-5 strength.
+    Identical,
+}
+
+impl From<CollationStrength> for u32 {
+    fn from(strength: CollationStrength) -> Self {
+        match strength {
+            CollationStrength::Primary => 1,
+            CollationStrength::Secondary => 2,
+            CollationStrength::Tertiary => 3,
+            CollationStrength::Quaternary => 4,
+            CollationStrength::Identical => 5,
+        }
+    }
+}
+
+impl TryFrom<u32> for CollationStrength {
+    type Error = Error;
+
+    fn try_from(level: u32) -> Result<Self, Self::Error> {
+        Ok(match level {
+            1 => CollationStrength::Primary,
+            2 => CollationStrength::Secondary,
+            3 => CollationStrength::Tertiary,
+            4 => CollationStrength::Quaternary,
+            5 => CollationStrength::Identical,
+            _ => {
+                return Err(ErrorKind::ArgumentError {
+                    message: (format!("invalid collation strength: {}", level)),
+                }
+                .into())
+            }
+        })
+    }
+}
+
+impl Serialize for CollationStrength {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let level = u32::from(*self);
+        serializer.serialize_i32(level as i32)
+    }
+}
+
+impl<'de> Deserialize<'de> for CollationStrength {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let level = u32::deserialize(deserializer)?;
+        Self::try_from(level).map_err(serde::de::Error::custom)
+    }
+}
+
+impl std::fmt::Display for CollationStrength {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(&u32::from(*self), f)
+    }
+}
+
+/// Setting that determines sort order of case differences during case tertiary level comparisons.
+/// For more info, see http://userguide.icu-project.org/collation/customization.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
```

Comment:
well TIL kebab case is a thing... does it actually do anything here though given these enum cases are all one word?